XCast/XCast.cpp

```diff
@@ -886,47 +889,47 @@ void XCast::getUrlFromAppLaunchParams (const char *app_name, const char *payload
     memset (url, '\0', url_len);
     if(strcmp(app_name,"YouTube") == 0) {
         if ((payload != NULL) && (additional_data_url != NULL)){
-            sprintf( url, "https://www.youtube.com/tv?%s&additionalDataUrl=%s", payload, additional_data_url);
+            snprintf( url, url_len, "https://www.youtube.com/tv?%s&additionalDataUrl=%s", payload, additional_data_url);
         }else if (payload != NULL){
-            sprintf( url, "https://www.youtube.com/tv?%s", payload);
+            snprintf( url, url_len, "https://www.youtube.com/tv?%s", payload);
         }else{
-            sprintf( url, "https://www.youtube.com/tv");
+            snprintf( url, url_len, "https://www.youtube.com/tv");
         }
     }
     else if(strcmp(app_name,"YouTubeTV") == 0) {
         if ((payload != NULL) && (additional_data_url != NULL)){
-            sprintf( url, "https://www.youtube.com/tv/upg?%s&additionalDataUrl=%s", payload, additional_data_url);
+            snprintf( url, url_len, "https://www.youtube.com/tv/upg?%s&additionalDataUrl=%s", payload, additional_data_url);
         }else if (payload != NULL){
-            sprintf( url, "https://www.youtube.com/tv/upg?%s", payload);
+            snprintf( url, url_len, "https://www.youtube.com/tv/upg?%s", payload);
         }else{
-            sprintf( url, "https://www.youtube.com/tv/upg?");
+            snprintf( url, url_len, "https://www.youtube.com/tv/upg?");
         }
     }
     else if(strcmp(app_name,"YouTubeKids") == 0) {
         if ((payload != NULL) && (additional_data_url != NULL)){
-            sprintf( url, "https://www.youtube.com/tv_kids?%s&additionalDataUrl=%s", payload, additional_data_url);
+            snprintf( url, url_len, "https://www.youtube.com/tv_kids?%s&additionalDataUrl=%s", payload, additional_data_url);
         }else if (payload != NULL){
-            sprintf( url, "https://www.youtube.com/tv_kids?%s", payload);
+            snprintf( url, url_len, "https://www.youtube.com/tv_kids?%s", payload);
         }else{
-            sprintf( url, "https://www.youtube.com/tv_kids?");
+            snprintf( url, url_len, "https://www.youtube.com/tv_kids?");
         }
     }
     else if(strcmp(app_name,"Netflix") == 0) {
         memset( url, 0, url_len );
-        strcat( url, "source_type=12" );
+        strncat( url, "source_type=12", url_len - strlen(url) - 1);
         if(payload != NULL)
         {
             const char * pUrlEncodedParams;
             pUrlEncodedParams = payload;
             if( pUrlEncodedParams ){
-                strcat( url, "&dial=");
-                strcat( url, pUrlEncodedParams );
+                strncat( url, "&dial=", url_len - strlen(url) - 1);
+                strncat( url, pUrlEncodedParams, url_len - strlen(url) - 1);
```

Comment:
Checking why we are doing url_len - strlen(url) - 1. It should be url_len - strlen(url). it's already memset to 0. Pls check other strncpy & strncat where we have added -1.

Reply:
For `strncat(dest, src, count);`
This function call does add a null terminator but it is not included in "count". So the function writes count number of characters to the buffer PLUS a null terminator. So there needs to be one spot left in the buffer so the function can append the null terminator after it reaches count.


For reference:
https://en.cppreference.com/w/c/string/byte/strncat#:~:text=The%20behavior%20is%20undefined%20if%20the%20destination%20array%20does%20not%20have%20enough%20space%20for%20the%20contents%20of%20both%20dest%20and%20the%20first%20count%20characters%20of%20src%2C%20plus%20the%20terminating%20null%20character.